File: packages/expo-brightness/android/src/main/java/expo/modules/brightness/BrightnessModule.java
package expo.modules.brightness;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.provider.Settings;
import android.view.WindowManager;

import org.unimodules.core.ExportedModule;
import org.unimodules.core.InvalidArgumentException;
import org.unimodules.core.ModuleRegistry;
import org.unimodules.core.Promise;
import org.unimodules.core.interfaces.ActivityProvider;
import org.unimodules.core.interfaces.ExpoMethod;
import org.unimodules.core.interfaces.ModuleRegistryConsumer;

public class BrightnessModule extends ExportedModule implements ModuleRegistryConsumer {
  private ModuleRegistry mModuleRegistry;

  public BrightnessModule(Context reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "ExpoBrightness";
  }

  @Override
  public void setModuleRegistry(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;
  }

  @ExpoMethod
  public void setBrightnessAsync(final float brightnessValue, final Promise promise) {
    final Activity activity = getCurrentActivity();
    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        try {
          WindowManager.LayoutParams lp = activity.getWindow().getAttributes();
          lp.screenBrightness = brightnessValue;
          activity.getWindow().setAttributes(lp); // must be done on UI thread
          promise.resolve(null);
        } catch (Exception e) {
          promise.reject("ERR_BRIGHTNESS", "Failed to set the current screen brightness", e);
        }
      }
    });
  }

  @ExpoMethod
  public void getBrightnessAsync(final Promise promise) {
    final Activity activity = getCurrentActivity();
    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        WindowManager.LayoutParams lp = getCurrentActivity().getWindow().getAttributes();
        if (lp.screenBrightness == WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE) {
          // system brightness is not overridden by the current activity, so just resolve with it
          getSystemBrightnessAsync(promise);
        } else {
          promise.resolve(lp.screenBrightness);
        }
      }
    });
  }

  @ExpoMethod
  public void getSystemBrightnessAsync(Promise promise) {
    try {
      int brightnessMode = Settings.System.getInt(
          getCurrentActivity().getContentResolver(),
          Settings.System.SCREEN_BRIGHTNESS_MODE
      );
      if (brightnessMode == Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC) {
        float brightness = Settings.System.getFloat(
            getCurrentActivity().getContentResolver(),
            // https://stackoverflow.com/questions/29349153/change-adaptive-brightness-level-programatically
            // this setting cannot be changed starting in targetSdkVersion 23, but it can still be read
            "screen_auto_brightness_adj"
        );
        promise.resolve((brightness + 1.0f) / 2);
      } else {
        String brightness = Settings.System.getString(
            getCurrentActivity().getContentResolver(),
            Settings.System.SCREEN_BRIGHTNESS
        );
        promise.resolve(Integer.parseInt(brightness) / 255f);
      }
    } catch (Exception e) {
      promise.reject("ERR_BRIGHTNESS_SYSTEM", "Failed to get the system brightness value", e);
    }
  }

  @ExpoMethod
  public void setSystemBrightnessAsync(float brightnessValue, Promise promise) {
    try {
      // we have to just check this every time
      // if we try to store a value for this permission, there is no way to know if the user has changed it
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(getCurrentActivity())) {
        promise.reject("ERR_BRIGHTNESS_PERMISSIONS_DENIED", "WRITE_SETTINGS permission has not been granted");
        return;
      }
      // manual mode must be set in order to change system brightness (sets the automatic mode off)
      Settings.System.putInt(
          getCurrentActivity().getContentResolver(),
          Settings.System.SCREEN_BRIGHTNESS_MODE,
          Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL
      );
      Settings.System.putInt(
          getCurrentActivity().getContentResolver(),
          Settings.System.SCREEN_BRIGHTNESS,
          Math.round(brightnessValue * 255)
      );
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject("ERR_BRIGHTNESS_SYSTEM", "Failed to set the system brightness value", e);
    }
  }

  @ExpoMethod
  public void useSystemBrightnessAsync(final Promise promise) {
    final Activity activity = getCurrentActivity();
    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        try {
          WindowManager.LayoutParams lp = activity.getWindow().getAttributes();
          lp.screenBrightness = WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE;
          activity.getWindow().setAttributes(lp); // must be done on UI thread
          promise.resolve(null);
        } catch (Exception e) {
          promise.reject("ERR_BRIGHTNESS", "Failed to set the brightness of the current screen", e);
        }
      }
    });
  }

  @ExpoMethod
  public void isUsingSystemBrightnessAsync(final Promise promise) {
    final Activity activity = getCurrentActivity();
    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        WindowManager.LayoutParams lp = getCurrentActivity().getWindow().getAttributes();
        promise.resolve(lp.screenBrightness == WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE);
      }
    });
  }

  @ExpoMethod
  public void getSystemBrightnessModeAsync(Promise promise) {
    try {
      int brightnessMode = Settings.System.getInt(
          getCurrentActivity().getContentResolver(),
          Settings.System.SCREEN_BRIGHTNESS_MODE
      );
      promise.resolve(brightnessModeNativeToJS(brightnessMode));
    } catch (Exception e) {
      promise.reject("ERR_BRIGHTNESS_MODE", "Failed to get the system brightness mode", e);
    }
  }

  @ExpoMethod
  public void setSystemBrightnessModeAsync(int brightnessMode, Promise promise) {
    try {
      // we have to just check this every time
      // if we try to store a value for this permission, there is no way to know if the user has changed it
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(getCurrentActivity())) {
        promise.reject("ERR_BRIGHTNESS_PERMISSIONS_DENIED", "WRITE_SETTINGS permission has not been granted");
        return;
      }
      Settings.System.putInt(
          getCurrentActivity().getContentResolver(),
          Settings.System.SCREEN_BRIGHTNESS_MODE,
          brightnessModeJSToNative(brightnessMode)
      );
      promise.resolve(null);
    } catch (InvalidArgumentException e) {
      promise.reject(e);
    } catch (Exception e) {
      promise.reject("ERR_BRIGHTNESS_MODE", "Failed to set the system brightness mode", e);
    }
  }

  private int brightnessModeNativeToJS(int nativeValue) {
    switch (nativeValue) {
      case Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC:
        return 1;
      case Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL:
        return 2;
      default:
        return 0;
    }
  }

  private int brightnessModeJSToNative(int jsValue) throws Exception {
    switch (jsValue) {
      case 1:
        return Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC;
      case 2:
        return Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL;
      default:
        throw new InvalidArgumentException("Unsupported brightness mode " + String.valueOf(jsValue));
    }
  }

  private Activity getCurrentActivity() {
    return mModuleRegistry.getModule(ActivityProvider.class).getCurrentActivity();
  }
}

File: packages/expo-brightness/android/src/main/java/expo/modules/brightness/BrightnessModule.kt
package expo.modules.brightness

import android.Manifest
import android.provider.Settings
import android.view.WindowManager
import expo.modules.core.errors.InvalidArgumentException
import expo.modules.interfaces.permissions.Permissions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.math.roundToInt

const val brightnessChangeEvent = "Expo.brightnessDidChange"

class BrightnessModule : Module() {
  private val currentActivity
    get() = appContext.throwingActivity

  override fun definition() = ModuleDefinition {
    Name("ExpoBrightness")

    // This is unused on Android. It is only here to suppress the native event emitter warning
    Events(brightnessChangeEvent)

    AsyncFunction("requestPermissionsAsync") { promise: Promise ->
      Permissions.askForPermissionsWithPermissionsManager(appContext.permissions, promise, Manifest.permission.WRITE_SETTINGS)
    }

    AsyncFunction("getPermissionsAsync") { promise: Promise ->
      Permissions.getPermissionsWithPermissionsManager(appContext.permissions, promise, Manifest.permission.WRITE_SETTINGS)
    }

    AsyncFunction("setBrightnessAsync") { brightnessValue: Float ->
      val lp = currentActivity.window.attributes
      lp.screenBrightness = brightnessValue
      currentActivity.window.attributes = lp // must be done on UI thread
    }.runOnQueue(Queues.MAIN)

    AsyncFunction<Float>("getBrightnessAsync") {
      val lp = currentActivity.window.attributes
      val brightness = if (lp.screenBrightness == WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE) {
        // system brightness is not overridden by the current activity, so just resolve with it
        getSystemBrightness()
      } else {
        lp.screenBrightness
      }

      return@AsyncFunction brightness
    }.runOnQueue(Queues.MAIN)

    AsyncFunction<Float>("getSystemBrightnessAsync") {
      return@AsyncFunction getSystemBrightness()
    }

    AsyncFunction("setSystemBrightnessAsync") { brightnessValue: Float ->
      // we have to just check this every time
      // if we try to store a value for this permission, there is no way to know if the user has changed it
      if (!Settings.System.canWrite(currentActivity)) {
        throw BrightnessPermissionsException()
      }
      // manual mode must be set in order to change system brightness (sets the automatic mode off)
      Settings.System.putInt(
        currentActivity.contentResolver,
        Settings.System.SCREEN_BRIGHTNESS_MODE,
        Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL
      )
      Settings.System.putInt(
        currentActivity.contentResolver,
        Settings.System.SCREEN_BRIGHTNESS,
        (brightnessValue * 255).roundToInt()
      )
    }

    AsyncFunction<Unit>("restoreSystemBrightnessAsync") {
      val lp = currentActivity.window.attributes
      lp.screenBrightness = WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE
      currentActivity.window.attributes = lp // must be done on UI thread
    }.runOnQueue(Queues.MAIN)

    AsyncFunction<Boolean>("isUsingSystemBrightnessAsync") {
      val lp = currentActivity.window.attributes
      return@AsyncFunction lp.screenBrightness == WindowManager.LayoutParams.BRIGHTNESS_OVERRIDE_NONE
    }.runOnQueue(Queues.MAIN)

    AsyncFunction<Int>("getSystemBrightnessModeAsync") {
      val brightnessMode = Settings.System.getInt(
        currentActivity.contentResolver,
        Settings.System.SCREEN_BRIGHTNESS_MODE
      )
      return@AsyncFunction brightnessModeNativeToJS(brightnessMode)
    }

    AsyncFunction("setSystemBrightnessModeAsync") { brightnessMode: Int ->
      // we have to just check this every time
      // if we try to store a value for this permission, there is no way to know if the user has changed it
      if (!Settings.System.canWrite(currentActivity)) {
        throw BrightnessPermissionsException()
      }
      Settings.System.putInt(
        currentActivity.contentResolver,
        Settings.System.SCREEN_BRIGHTNESS_MODE,
        brightnessModeJSToNative(brightnessMode)
      )
    }
  }

  private fun getSystemBrightness(): Float {
    val brightnessMode = Settings.System.getInt(
      currentActivity.contentResolver,
      Settings.System.SCREEN_BRIGHTNESS_MODE
    )
    return if (brightnessMode == Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC) {
      val brightness = Settings.System.getFloat(
        currentActivity.contentResolver, // https://stackoverflow.com/questions/29349153/change-adaptive-brightness-level-programatically
        // this setting cannot be changed starting in targetSdkVersion 23, but it can still be read
        "screen_auto_brightness_adj"
      )
      (brightness + 1.0f) / 2
    } else {
      val brightness = Settings.System.getString(
        currentActivity.contentResolver,
        Settings.System.SCREEN_BRIGHTNESS
      )
      brightness.toInt() / 255f
    }
  }

  private fun brightnessModeNativeToJS(nativeValue: Int): Int {
    return when (nativeValue) {
      Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC -> 1
      Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL -> 2
      else -> 0
    }
  }

  @Throws(InvalidArgumentException::class)
  private fun brightnessModeJSToNative(jsValue: Int): Int {
    return when (jsValue) {
      1 -> Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC
      2 -> Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL
      else -> throw InvalidArgumentException("Unsupported brightness mode $jsValue")
    }
  }
}
